Update setConsent to react-spectrum3, fixes to RadioGroupWithDataElement

jonsnyder opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All tests pass and I've made any necessary test changes.
- [ ] I've updated the schema in extension.json or no changes are necessary.

jonsnyder commented on src/view/components/rawDataElementSelector.styl:
```diff
@@ -11,5 +11,5 @@ governing permissions and limitations under the License.
 
 .DataElementSelector-buttonAdjustedForLabel {
-  margin-top: 24px;
+  margin-top: 24px !important;
```
I don't understand why this is necessary.

Aaronius replied:
I dug into this. It turns out the styles from `@adobe/react-spectrum` are bundled in our CSS file before `.DataElementSelector-buttonAdjustedForLabel`, which is good, but the styles from `@react-spectrum/combobox` are bundled into the CSS file _after_ `.DataElementSelector-buttonAdjustedForLabel`. The combobox styles include a `._spectrum-ActionButton_7a745,._spectrum-Button_7a745,._spectrum-ClearButton_7a745,._spectrum-FieldButton_7a745,._spectrum-LogicButton_7a745` selector that sets margin to 0. Since this selector comes after `.DataElementSelector-buttonAdjustedForLabel`, its 0 margin wins.

I don't think there's a great way of forcing the ComboBox's styles to be bundled into the CSS file before our own styles (I've asked on the react-spectrum channel. I'll let you know if I hear otherwise.). Considering this, using `!important` here is a reasonable workaround, but I think it would be better to add a `marginTop="size-300"` prop to the button inside `rawDataElementSelector.jsx` instead of having this class. I probably should have done that in the first place.

jonsnyder commented on src/view/components/formikReactSpectrum3/radioGroupWithDataElement.jsx:
```diff
@@ -75,14 +75,17 @@ const RadioGroupWithDataElement = ({
   });
```
The setConsent tests found some bugs in the radioGroupWithDataElement component.

Aaronius commented:
If I load the view using this for settings:

```
{
    "instanceName": "alloy",
    "identityMap": "%dataElement9865%",
    "consent": "%dataElement2sdf861%"
  }
```

and then switch from "use a data element" to "fill out a form", I get an error caught by the error boundary.

![image](https://user-images.githubusercontent.com/210820/123444624-51d93e80-d594-11eb-89f3-81b5ba66421f.png)

Aaronius commented:
When I load the view using this for settings:

```
{
  "instanceName": "alloy",
  "consent": [
    {
      "standard": "Adobe",
      "version": "1.0",
      "value": {
        "general": "%dataElement9680%"
      }
    },
    {
      "standard": "Adobe",
      "version": "2.0",
      "value": "%dataElement8346%"
    },
    {
      "standard": "Adobe",
      "version": "1.0",
      "value": {
        "general": "in"
      }
    }
  ]
}
```

and I switch the first consent from "use a data element" to "in" and then back to "use a data element", the value that was in "use a data element" is gone, which I believe we were trying to avoid.

jonsnyder commented:
@Aaronius thank you for taking the time to test it. I've fixed that error and added an automated test for it.

Aaronius commented on src/view/components/formikReactSpectrum3/radioGroupWithDataElement.jsx:
```diff
@@ -77,18 +77,26 @@ const RadioGroupWithDataElement = ({
   }
 
   const setValues = (newRadioValue, newDataElement) => {
+    console.log("setValues", newRadioValue, newDataElement);
```
Remove.

jonsnyder commented:
@Aaronius Thanks for finding these two bugs!  I fixed them and addressed your comments. I'm not sure about the spacing issue. Let's meet on Monday to discuss the spacing.

Aaronius commented on src/view/components/rawDataElementSelector.jsx:
```diff
@@ -45,9 +44,7 @@ const RawDataElementSelector = ({
-        UNSAFE_className={
-          adjustForLabel ? "DataElementSelector-buttonAdjustedForLabel" : ""
-        }
+        UNSAFE_style={adjustForLabel ? { marginTop: "size-300" } : {}}
```
This won't work because `size-300` isn't a valid CSS value. You'll need to use the `marginTop` prop. It's a prop provided by React-Spectrum that does the interpretation from `size-300` to 24px.
